src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -168,28 +168,108 @@ internal bool TryAddWithoutValidation(HeaderDescriptor descriptor, string? value
         internal bool TryAddWithoutValidation(HeaderDescriptor descriptor, IEnumerable<string?> values)
         {
             ArgumentNullException.ThrowIfNull(values);
+            string?[]? valuesArray = values as string?[];
+            IList<string?>? valuesList = values as IList<string?>;
```

Comment:
This is now doing two type checks on the hot path.

I would only check for `IList` upfront. If it turns out that the count is > 1, you could always do the type test for `string[]` before the loop, and keep the previous code simpler